### PR TITLE
ControlNet support

### PR DESCRIPTION
**Pull request by brianfitzgerald**

### CODE REVIEW

**Review by voletiv — CHANGES_REQUESTED** [review]

File: other_impls.py [hunk 1]
```diff
@@ -7,21 +7,33 @@
...
-    out = torch.nn.functional.scaled_dot_product_attention(
-        q, k, v, attn_mask=mask, dropout_p=0.0, is_causal=False
-    )
-    return out.transpose(1, 2).reshape(b, -1, heads * dim_head)
+    if attn_mode == "torch":
+        out = torch.nn.functional.scaled_dot_product_attention(
+            q, k, v, attn_mask=mask, dropout_p=0.0, is_causal=False
+        )
+        return out.transpose(1, 2).reshape(b, -1, heads * dim_head)
+    elif attn_mode == "xformers":
```
[the remark this answers is not in the document]
Comment (brianfitzgerald) [reply other_impls.py:33]: Done, I've removed some straggler `attn_mode` references that were previously in the code.

File: README.md [hunk 2]
```diff
@@ -47,10 +47,13 @@ python3 sd3_infer.py --prompt path/to/my_prompts.txt --model models/sd3.5_medium
...
-To add a postfix to the output directory, add `--postfix <my_postfix>`. For example,
 ```sh
-python3 sd3_infer.py --prompt path/to/my_prompts.txt --postfix "steps100" --steps 100
+# Generate a cat on ref model with default settings
+python3 -s sd3_infer.py
+# Generate a 1024 cat on SD3-8B
+python3 -s sd3_infer.py --width 1024 --height 1024 --shift 3 --model models/sd3_medium.safetensors --prompt "cute wallpaper art of a cat"
+# Or for parameter listing
+python3 -s sd3_infer.py --help
 ```
```
[the remark this answers is not in the document]
Comment (brianfitzgerald) [reply README.md]: added

File: sd3_infer.py [hunk 3]
```diff
@@ -294,21 +382,29 @@ def do_sampling(
...
         self.print("Encoding image to latent...")
         image = image.convert("RGB")
         image_np = np.array(image).astype(np.float32) / 255.0
         image_np = np.moveaxis(image_np, 2, 0)
         batch_images = np.expand_dims(image_np, axis=0).repeat(1, axis=0)
-        image_torch = torch.from_numpy(batch_images)
-        image_torch = 2.0 * image_torch - 1.0
+        image_torch = torch.from_numpy(batch_images).cuda()
+        if using_8b_controlnet:
+            image_torch = 2.0 * image_torch - 1.0
```
Comment (voletiv) [inline sd3_infer.py:393]: I think this part is wrong?
This line is supposed to be within the non-controlnet case (as well?), right?

File: sd3_infer.py [hunk 4]
```diff
@@ -111,34 +125,79 @@ def __init__(self):
...
+            if control_model_file is not None:
+                control_model_ckpt = safe_open(
+                    control_model_file, framework="pt", device=device
+                )
             self.model = BaseModel(
                 shift=shift,
                 file=f,
                 prefix="model.diffusion_model.",
-                device="cpu",
+                device="cuda",
```
Comment (voletiv) [inline sd3_infer.py:169]: The model is later loaded into cuda here : https://github.com/Stability-AI/sd3.5/blob/4e484e05308d83fb77ae6f680028e6c313f9da54/sd3_infer.py#L270, should this be cuda here?

File: sd3_infer.py [hunk 5]
```diff
@@ -171,42 +230,65 @@ def __init__(self, model):
...
+        if load_tokenizers:
+            print("Loading Google T5-v1-XXL...")
+            self.t5xxl = T5XXL(model_folder, text_encoder_device, torch.float32)
+            print("Loading OpenAI CLIP L...")
+            self.clip_l = ClipL(model_folder)
+            print("Loading OpenCLIP bigG...")
+            self.clip_g = ClipG(model_folder, text_encoder_device)
         print(f"Loading SD3 model {os.path.basename(model)}...")
-        self.sd3 = SD3(model, shift, verbose)
+        self.sd3 = SD3(model, shift, controlnet_ckpt, verbose, "cuda")
```
Comment (voletiv) [inline sd3_infer.py:279]: Same comment as earlier. Models are to be loaded into and offloaded from GPU just before and after they are used, right?

**Review by voletiv — CHANGES_REQUESTED** [review]

File: sd3_infer.py (hunk 4 above)
Comment (brianfitzgerald) [reply sd3_infer.py:169]: I think we want the dit to always be in cuda memory no? The text encoders we want to offload to `text_encoder_device`, but the diffusion model should stay in memory until program end

**Pull request merged**